### PR TITLE
add metrics driver for Huawei OceanStor

### PR DESCRIPTION
**Pull request by xxwjj**
**What this PR does / why we need it**:
1. add metrics driver for Huawei OceanStor
**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

File: contrib/drivers/huawei/dorado/metrics.go [hunk 1]
```diff
@@ -0,0 +1,232 @@
+// Copyright (c) 2019 The OpenSDS Authors.
+//
+//    Licensed under the Apache License, Version 2.0 (the "License"); you may
```
Comment (leonwanghui) [inline contrib/drivers/huawei/dorado/metrics.go]: This style of copyright looks a bit odd, could u take a look at it?

File: contrib/drivers/huawei/dorado/metrics.go [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+var data = `
+resources:
+  - resource: volume
+    metrics:
+      - iops
+      - read_throughput
+      - write_throughput
+      - response_time
+      - service_time
+      - utilization_prcnt
```
Comment (xing-yang) [inline contrib/drivers/huawei/dorado/metrics.go]: Can you add comments on the naming of these metrics on the storage system if they are called something else?

Comment (xing-yang) [reply contrib/drivers/huawei/dorado/metrics.go]: Add a "TODO" to add resources for "disk" and "port".

File: contrib/drivers/huawei/dorado/metrics.go [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+    units:
+      - tps
+      - kbs
+      - kbs
+      - ms
+      - ms
+      - '%'
```
Comment (NajmudheenCT) [inline contrib/drivers/huawei/dorado/metrics.go]: Change '%' to 'prcnt' , exporter/push mechanism has issue in parsing '%. refer https://github.com/opensds/opensds/issues/787
I will making changes for this in lvm driver also.

File: contrib/drivers/huawei/dorado/metrics.go [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+      - response_time
+      - service_time
+      - utilization_prcnt
+    units:
+      - tps
+      - kbs
+      - kbs
+      - ms
+      - ms
+      - '%'`
```
Comment (NajmudheenCT) [inline contrib/drivers/huawei/dorado/metrics.go]: same here

File: contrib/drivers/huawei/dorado/metrics.go [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+      - response_time
+      - service_time
+      - utilization_prcnt
+    units:
+      - tps
+      - kbs
+      - kbs
+      - ms
+      - ms
+      - '%'
```
Comment (NajmudheenCT) [inline contrib/drivers/huawei/dorado/metrics.go]: same here

File: contrib/drivers/huawei/dorado/metrics.go [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+				Timestamp: getCurrentUnixTimestamp(),
+				Value:     val,
+			}
+			metricValues := make([]*model.Metric, 0)
+			metricValues = append(metricValues, metricValue)
+			metric := &model.MetricSpec{
+				InstanceID:   uuid,
+				InstanceName: uuid,
+				Job:          "HuaweiOceanStor",
+				Labels:       map[string]string{},
```
Comment (NajmudheenCT) [inline contrib/drivers/huawei/dorado/metrics.go]: Please add some important properties related to the metrics in to label map. at least add  component name.

File: contrib/drivers/huawei/dorado/metrics.go (hunk 3 above)
Comment (xxwjj) [reply contrib/drivers/huawei/dorado/metrics.go]: ok, thanks

File: contrib/drivers/huawei/dorado/metrics.go [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+				Timestamp: getCurrentUnixTimestamp(),
+				Value:     metricMap[element],
+			}
+			metricValues := make([]*model.Metric, 0)
+			metricValues = append(metricValues, metricValue)
+			metric := &model.MetricSpec{
+				InstanceID:   uuid,
+				InstanceName: uuid,
+				Job:          "HuaweiOceanStor",
+				Labels:       map[string]string{"device": MetricResourceTypeController},
```
Comment (NajmudheenCT) [inline contrib/drivers/huawei/dorado/metrics.go]: I suggest controller ID as the Label
Labels:       map[string]string{"controller": uuid}

File: contrib/drivers/huawei/dorado/metrics.go [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+				Timestamp: getCurrentUnixTimestamp(),
+				Value:     metricMap[element],
+			}
+			metricValues := make([]*model.Metric, 0)
+			metricValues = append(metricValues, metricValue)
+			metric := &model.MetricSpec{
+				InstanceID:   poolId,
+				InstanceName: pool.Name,
+				Job:          "HuaweiOceanStor",
+				Labels:       map[string]string{"device": MetricResourceTypePool},
```
Comment (NajmudheenCT) [inline contrib/drivers/huawei/dorado/metrics.go]: I suggest poolid as the lable here 
Labels:       map[string]string{"pool": poolId},

File: contrib/drivers/huawei/dorado/metrics.go [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dorado
+
+import (
+	"github.com/satori/go.uuid"
```
Comment (leonwanghui) [inline contrib/drivers/huawei/dorado/metrics.go]: Please move this path.

File: contrib/drivers/huawei/dorado/metrics.go [hunk 10]
```diff
@@ -0,0 +1,311 @@
...
+      - latency
+      - service_time
+      - cache_hit_ratio
+      - cpu_usage
+    units:
+      - tps
+      - mbs
+      - microsecond
+      - ms
+      - prcnt
```
Comment (leonwanghui) [inline contrib/drivers/huawei/dorado/metrics.go]: I think it's duplicated here.

File: contrib/drivers/huawei/dorado/metrics.go [hunk 11]
```diff
@@ -0,0 +1,311 @@
...
+		path = defaultConfPath
+	}
+
+	conf := &DoradoConfig{}
+	Parse(conf, path)
+
+	d.conf = conf
+	d.client, err = NewClient(&d.conf.AuthOptions)
+	if err != nil {
+		log.Errorf("Get new client failed, %v", err)
```
Comment (leonwanghui) [inline contrib/drivers/huawei/dorado/metrics.go]: s/Get/get
s/,/:

File: contrib/drivers/huawei/dorado/metrics.go (hunk 10 above)
Comment (xxwjj) [reply contrib/drivers/huawei/dorado/metrics.go]: no, one for controller and one for pool , they are not the same

File: contrib/drivers/huawei/dorado/metrics.go (hunk 7 above)
Comment (xxwjj) [reply contrib/drivers/huawei/dorado/metrics.go]: @NajmudheenCT done, BTW, can you tell me what is Lables used for?

Comment (NajmudheenCT) [reply contrib/drivers/huawei/dorado/metrics.go]: labels are to differentiate the characteristics of the thing that is being measured, That could be associated components/properties like controller,pool,port,storageGroup, etc..

**Pull request merged**